### PR TITLE
TradeShips hyperjump code change

### PR DESCRIPTION
**Pull request by Max5377**
This PR is an attempt to fix cluttering hyperclouds around stations and rare illegal jumps almost at the planet's surface. Now `TradeShips` behaviour when undocking looks like this.
For orbital stations:
1. Undock;
2. Fly to limits of current orbital station;
3. `OnAICompleted` event fires, tell ship to fly to current system's star;
4. Assign task to check distance between current orbital station and ship, if distance is large enough, make the jump.

For ground stations;
1. Undock;
2. Tell ship to fly to current system's star;
3. Assign task to check altitude between current orbital station's parent and ship, if altitude is high enough, make the jump.

The distance and altitude will be selected from a random interval of 20,000 - 240,000 meters.
Demonstration:
1. Undocking and jumping from Cydonia

https://github.com/pioneerspacesim/pioneer/assets/69468517/bd73a4d7-0f1f-4607-bc64-b5e843707223

2. Undocking and jumping from Gates Starport
   

https://github.com/pioneerspacesim/pioneer/assets/69468517/eabefebd-fcd8-4e99-8a70-1ec8fba50b99

**Update**: rebase.
**Update2**: `addEquip` is removed from `setPlayerAsTraderDocked` and `setPlayerAsTraderInbound`.
Added check if player is currently docked (`setPlayerAsTraderDocked`) or there's nearest station (`setPlayerAsTraderInbound`).
If not, player will not become trader.
**Update3**: rebase.

### CODE REVIEW

**Comment by Max5377** [comment]
There's still a possibility for this(ex. for Andersen City in Ahonen's Claim in sector Lalande 21185):
Far view:
AndersenCityCloudsFar

Station view:
AndersenCityCloudsStation

But in my opinion it doesn't look this bad as showed here:
https://github.com/pioneerspacesim/pioneer/issues/5539#issuecomment-1437330530
https://github.com/pioneerspacesim/pioneer/issues/5539#issuecomment-1449327723

**Comment by Max5377** [comment]
@Gliese852 
1. There's a possibility of inbound ship coliding with planet's terrain, exploding in the process. Maybe it's better to tell ship to orbit planet first and only then dock with station?
   (autopilot also has this problem);
2. There is still a possibility of undocking ship breaking the rules and making illegal hyperjump when it's under attack
   (Perhaps can be explained that pilot wants to save his live, not caring about rules at the moment);
3. You were right. Ships are jumping at higher distances when game speed is 10000x. I also discovered this behaviour in the original `TradeShips` code too
   (Can't think of any solution for this, unfortenately, unless it's not big of a deal);
4. The second step of behaviour in orbital stations would be skipped if there wasn't a problem with pathfinding when ship is inside orbital station
   (They basically will ignore walls, start to collide with them and possibly exploding in the process);
5. I included `setPlayerAsTraderDocked` and `setPlayerAsTraderInbound` in `Flow` part of `TradeShips` so you can view `TradeShips` AI from player's eyes (used in the demonstration)
   (Let me know if you want to move it from here or delete it completely);
6. Also, right now in my changed code I'm using `GetAltitudeRelTo`, which has a problem of using player pos in calculating altitude, skipping first argument, but I've done [PR](https://github.com/pioneerspacesim/pioneer/pull/5643) that I mentioned above.
Aside from this I haven't discovered any problems.

**Comment by Max5377** [comment]
Forgot to mention that I also had rare case with old code, where when I was jumping to Bernard's Star from Sol and back and going to the Moon after that, I would see hyperspace cloud below planet's terrain. I have a save to demonstrate what I'm talking about:
[CloudBelowTerrain.zip](https://github.com/pioneerspacesim/pioneer/files/12903463/CloudBelowTerrain.zip)

Haven't seen this with my change, but who knows. I hope this will fix this.

**Comment by Gliese852** [comment]
@Max5377 
> 1. There's a possibility of inbound ship coliding with planet's terrain, exploding in the process. Maybe it's better to tell ship to orbit planet first and only then dock with station?
> (autopilot also has this problem);

Ideally, of course, it would be worth solving this problem inside the autopilot, but if such a workaround will reduce the number of victims, I’m for it.

> 2. There is still a possibility of undocking ship breaking the rules and making illegal hyperjump when it's under attack
> (Perhaps can be explained that pilot wants to save his live, not caring about rules at the moment);

I think this is quite correct behavior.

> 3. You were right. Ships are jumping at higher distances when game speed is 10000x. I also discovered this behaviour in the original `TradeShips` code too
> (Can't think of any solution for this, unfortenately, unless it's not big of a deal);

I'm inclined to think that this is not big of a deal.

> 4. The second step of behaviour in orbital stations would be skipped if there wasn't a problem with pathfinding when ship is inside orbital station
> (They basically will ignore walls, start to collide with them and possibly exploding in the process);

It’s strange, it works the same for me, besides, both commands ultimately come down to calling `AICmdFlyTo`, where it turns out that the body is too close to the center of the orbital and the ship tends [to move away normally](https://github.com/pioneerspacesim/pioneer/blob/00b87acd037d14b46db3ffacc14711c7de854686/src/ShipAICmd.cpp#L1112). Maybe you could provide a save file where the difference in behavior is reproduced?

> 5. I included `setPlayerAsTraderDocked` and `setPlayerAsTraderInbound` in `Flow` part of `TradeShips` so you can view `TradeShips` AI from player's eyes (used in the demonstration)
> (Let me know if you want to move it from here or delete it completely);

There are some problems with these functions, as seen in the second video, Probably, you shouldn’t try to `Trader.addEquip`, because it falls with assertion. Overall this seems like it might be quite useful. I would also like to note that undocking on the autopilot already works for the player, so you can give the autopilot a command while on the pad and see what happens.

**Comment by Max5377** [comment]
@Gliese852 
> Ideally, of course, it would be worth solving this problem inside the autopilot, but if such a workaround will reduce the number of victims, I’m for it.

Didn't do anything really. The death rate for `Sol` system is ~50+/-10 ships per week no matter what approach is used.
Also wanted to add that ships are sometimes colliding with each other too (inbound and outbound) and looks like it usually happenes at orbital stations (according to `TradeShips` log). 
Let `Lalande 21185` be a stress-test for fixing ships AI, because the death rate for this system is insane (179 ships in a week first try) and for some reason in this system there's a lot of deaths from the star itself (according to `TradeShips` log too).

> It’s strange, it works the same for me, besides, both commands ultimately come down to calling AICmdFlyTo, where it turns out that the body is too close to the center of the orbital and the ship tends [to move away normally](https://github.com/pioneerspacesim/pioneer/blob/00b87acd037d14b46db3ffacc14711c7de854686/src/ShipAICmd.cpp#L1112). Maybe you could provide a save file where the difference in behavior is reproduced?

My bad. I'll try to show and explain what happenes. And it's completely different for me.

Giving ship command to fly to body, when ship is more to the center of orbital starport, the more chance it will deviate from the course and fly into the wall instead, especially if it rotates backwards beforehand. 
Here's video to demonstrate this:

https://github.com/pioneerspacesim/pioneer/assets/69468517/244e6b91-e981-424a-a09e-db40e5cda600


If ship is in the center of orbital station, he will usually fly upwards and also try to go through wall, no matter if it's moving or not.
Demonstration:

https://github.com/pioneerspacesim/pioneer/assets/69468517/345d3768-4cc0-481d-a660-8bc5b210053c

This happenes with all commands that tell ship to fly to something (`AIFlyTo`, `AIEnterLow`/`Medium`/`HighOrbit`).
You can also experience it by going into big stations (`Jobs Pad`, `Gates Spaceport`) and setting maximum speed. Soon you will hear collision sounds. I will give you the save, but...

I spotted another issue:
1. Start new game;
2. Wait around for some time;
3. Save;
4. Load;
5. Wait around for some time.

There's chance of game failing with this:
AssertionFailed

After diging around more, this happenes at this line in `Events.lua` in `onShipDocked`:

```
local damage = ShipDef[trader.ship_name].hullMass - ship.hullMassLeft
if damage > 0 then
	ship:SetHullPercent()
	Trader.addEquip(ship) <--
end
```

This behaviour is also reproducable in the original `TradeShips` code.

[Ship's behaviour on Orbital stations.zip](https://github.com/pioneerspacesim/pioneer/files/12908995/Ship.s.behaviour.on.Orbital.stations.zip)

> There are some problems with these functions, as seen in the second video, Probably, you shouldn’t try to Trader.addEquip, because it falls with assertion. Overall this seems like it might be quite useful. I would also like to note that undocking on the autopilot already works for the player, so you can give the autopilot a command while on the pad and see what happens.

Yes. It is enough to order the ship to undock after making the player a trader. No point in exposing `trader_task` table in `Trader.lua` (I didn't include this anyway, but showed it in the demonstration). I will also delete `addEquip`. And I don't know about `Space.PutShipOnRoute`. It frequently puts my ship in what looks like the center of the ground station's model, unless I haven't forgot something (or it's intended).
Demonstration (I set `Engine.rand:Number(0.0, 0.999)` to `Engine.rand:Number(0.999, 0.999)` beforehand):
For ground station:

https://github.com/pioneerspacesim/pioneer/assets/69468517/0589ac9f-b83b-41bf-ba43-37e2bc690940

For orbital station:

https://github.com/pioneerspacesim/pioneer/assets/69468517/13b8de9d-901f-460d-a497-d37c145a962f

**Comment by Gliese852** [comment]
@Max5377 have you read the [description of PutShipOnRoute](https://github.com/pioneerspacesim/pioneer/blob/04c611141b623d26ac155cd3f2708756f40101ab/src/lua/LuaSpace.cpp#L194)? I would like to point out that if the distance to the target is short enough, a `t_ratio` of `0.999` will place you almost at the target coordinates.
The only purpose of this function is to arrange the tradeships in the system when the player enters the star system (or starts the new game) as if they had been flying here for a long time already.

>This happenes with all commands that tell ship to fly to something (AIFlyTo, AIEnterLow/Medium/HighOrbit).

Yes, because, as I already wrote, the first thing a ship does is try to get out of the "object’s radius" _along the normal_, that is, along the ray connecting the center of the station and the center of the ship. So, if the ship is lucky and it is exactly on the axis of the station, it will fly through the gate, if it is shifted off the axis, it will crash somewhere.

**Comment by Max5377** [comment]
@Gliese852 Sorry, math is just not my best subject, really. But, I kinda understand what you mean now. Thanks for clarification.
Do you have any suggestions about `TradeShips` code itself or all good?

**Comment by Gliese852** [comment]
@Max5377 Looks like it's still a draft? I see blots in the code (occasionally spaces/tabs at the end of the line), commented out pieces.

**Comment by Max5377** [comment]
@Gliese852 I deleted parts of commented old code that doesn't needed anymore.
Apart from the issue that I mentioned [here](https://github.com/pioneerspacesim/pioneer/pull/5647#issuecomment-1762137010) about `GetAltitudeRelTo`, it should work without any other issues.

File: data/modules/TradeShips/Events.lua
Comment (Gliese852) [inline data/modules/TradeShips/Events.lua:149]: It’s still unclear, is there any point in this stage?

File: data/modules/TradeShips/Events.lua
Comment (Gliese852) [inline data/modules/TradeShips/Events.lua:164]: There is an extra space at the end of the line.

File: data/modules/TradeShips/Flow.lua
Comment (Gliese852) [inline data/modules/TradeShips/Flow.lua:511]: This function should currently be used immediately after exiting hyperspace - when the ship is in orbit around a star. I think it's worth either teleporting the player to the star, or writing a comment about the applicability of `setPlayerAsTraderInbound`.

File: data/modules/TradeShips/Trader.lua
Comment (Gliese852) [inline data/modules/TradeShips/Trader.lua:283]: might work `stationsParent = trader.starport.frameBody.body`

File: data/modules/TradeShips/Trader.lua
Comment (Gliese852) [inline data/modules/TradeShips/Trader.lua:313]: This is not super important IMHO, but you can get rid of 2 indentations if you exit using the opposite condition, i.e. `if not trader then return end`

File: data/modules/TradeShips/Events.lua
Comment (Gliese852) [inline data/modules/TradeShips/Events.lua:61]: Please take a closer look; in many places you have tabs on an empty line - it would be nice to not allow this.

File: data/modules/TradeShips/Events.lua
Comment (Max5377) [reply data/modules/TradeShips/Events.lua:149]: There's a high chance of ships crashing into walls if given the command to fly to star directly when they are inside station. Let this be for now. This command will still be given, when `AIFlyTo` get completed (~15km from station).

File: data/modules/TradeShips/Trader.lua
Comment (Max5377) [reply data/modules/TradeShips/Trader.lua:283]: `frameBody` only works for dynamic bodies. Station is not dynamic body, sadly.
`LuaBody.cpp`:
```
/*
 * Attribute: frameBody
 *
 * The non-dynamic body attached to the frame this dynamic body is in.
 *
 * Only valid for dynamic <Bodies>. For non-dynamic bodies <frameBody> will be
 * nil.
 ...
```

File: data/modules/TradeShips/Events.lua
Comment (Max5377) [reply data/modules/TradeShips/Events.lua:61]: Ok, I will try to do this on clear head, to not miss anything.

File: data/modules/TradeShips/Flow.lua
Comment (Max5377) [reply data/modules/TradeShips/Flow.lua:511]: I will try to teleport player to star (or add comment if I fail), but what's so strange in current behaviour to do this step?

File: data/modules/TradeShips/Events.lua
Comment (Gliese852) [reply data/modules/TradeShips/Events.lua:149]: It would be good to see confirmation of this statement, as I already wrote, this does not reproduce for me - you have the same chance of leaving the orbital no matter what autopilot command you run.

File: data/modules/TradeShips/Flow.lua
Comment (Gliese852) [reply data/modules/TradeShips/Flow.lua:511]: If I understood the question correctly, for example, you described some strange behavior:
> And I don't know about Space.PutShipOnRoute. It frequently puts my ship in what looks like the center of the ground station's model, unless I haven't forgot something (or it's intended).

Comment (Max5377) [reply data/modules/TradeShips/Flow.lua:511]: Ah. I readed description for `Space.PutShipOnRoute` more closely and it says:
```
/*
 * Function: PutShipOnRoute
 *
 * The ship rearranged from its current position to a given body in space, for a
 * given part of the path, as if it were flying in a straight line, consuming
 * fuel and changing speed.
```
Basically, if `t_ratio` == 0.999, ship teleports to station's icon as you can see on `PutShipAtPointOfStation_Ground.mp4` video. This happened for me more frequently, because path between ship and station was short and as it says in the description:
**`The ship rearranged from its current position to a given body in space`**
But I already wrote `Space.PutShipIntoOrbit` on my computer, so I'm gonna include it in the next `force-push`.

File: data/modules/TradeShips/Events.lua
Comment (Max5377) [reply data/modules/TradeShips/Events.lua:149]: There is not much to add, besides what I have already described in this [comment](https://github.com/pioneerspacesim/pioneer/pull/5647#issuecomment-1763326176). Try to fly to the center of `Jobes Pad` (area between landings pads) and give ship, for example, command to fly to current station. There's almost a guarantee ship will not fly to the exit, but will fly upwards to the wall. Or just give this command, when exit is behind ship and it's moving from it. The higher the speed of the ship, the higher the probability that it will simply fly to the wall, not upwards. 
These two behaviors will occur even if there is no obstructor on the way out. I can reproduce this on current release, even on current master.

File: data/modules/TradeShips/Events.lua
Comment (Max5377) [reply data/modules/TradeShips/Events.lua]: Not anymore.

File: data/modules/TradeShips/Trader.lua
Comment (Max5377) [reply data/modules/TradeShips/Trader.lua]: Changed to this and also changed the way to get first system's star to:
`trader.starport:GetSystemBody().system:GetStars()[1].body`

File: data/modules/TradeShips/Trader.lua
Comment (Max5377) [reply data/modules/TradeShips/Trader.lua]: Done

File: data/modules/TradeShips/Events.lua
Comment (Max5377) [reply data/modules/TradeShips/Events.lua]: There should be no more such errors.
Thanks to @impaktor for suggesting to install plugin for `editorconfig`-file.
P.S.: I use `Notepad++` for `lua` code. This also has built-in feature to trim trailing spaces on save.

File: data/modules/TradeShips/Flow.lua
Comment (Max5377) [reply data/modules/TradeShips/Flow.lua]: Added `PutShipIntoOrbit` in `LuaSpace.cpp`. Using this in the `Flow.setPlayerAsTraderInbound`.

**Comment by Max5377** [comment]
@Gliese852 Also pushed fix for `lua` error when you try to hyperspace with `TradeShips` debug tab opened.

File: data/modules/TradeShips/Events.lua
Comment (Gliese852) [reply data/modules/TradeShips/Events.lua:147]: Please explain why do you give the `AIFlyTo(starport)` command if
> ... give ship, for example, command to fly to current station. There's almost a guarantee ship will not fly to the exit, but will fly upwards to the wall.

Comment (Max5377) [reply data/modules/TradeShips/Events.lua]: Decided to try giving command to fly to low orbit of first system's star immediatly when ship undocks once more and I haven't seen much difference in how ship behaves from previous choice, so I pushed change so now ship will do exactly this. When I first tested this, it seemed to me that with this choice everything just gotten worse, so I decided to quickly revert to `AIFlyTo`.
At this point, let's just consider this issues unrelated to `TradeShips` code. Separate issues can be opened for this if needed.

**Review by Web-eWorks — CHANGES_REQUESTED** [review]
I'll generally leave the approval and merge duties to @Gliese852 as this is code he's most familiar with. I apologize for the long delay in contributing my review of this code - I've only noticed one issue that I'd appreciate being addressed before merge, but I otherwise would like to merge this in the next few weeks if there are no outstanding objections.

Thank you for addressing these bugs!

**Comment by Gliese852** [comment]
@Max5377  I think it’s worth doing a rebase again, in the current state I always have a crash at the beginning of the hyperjump.

**Comment by Max5377** [comment]
@Gliese852 Did rebase to current master. Shouldn't crash now.

File: data/modules/TradeShips/Debug.lua
Comment (Gliese852) [inline data/modules/TradeShips/Debug.lua:108]: One more point - maybe even disable the entire tab during hyperspace? iI seems that nothing but the log is relevant there anymore (and the log is not particularly relevant).

Comment (Max5377) [reply data/modules/TradeShips/Debug.lua:108]: I think there is no harm if the tab stays during hyperspace. It's short sequence anyway. Maybe log should be cleared when player is in hyperspace, because ships that are there are deleted by that point?

Comment (Gliese852) [reply data/modules/TradeShips/Debug.lua:108]: How to say, for example, a flight on Coronatrix 36 light years lasts 45 seconds of real time, enough to go to 'Local routes', click somewhere and crash the game. Do we have a good reason to show this tab in hyperspace?

Comment (Max5377) [reply data/modules/TradeShips/Debug.lua]: `TradeShips` debug tab now will be disabled in the hyperspace.

Comment (Gliese852) [reply data/modules/TradeShips/Debug.lua]: Thanks!

File: src/lua/LuaSpace.cpp
Comment (Gliese852) [reply src/lua/LuaSpace.cpp]: @Max5377 looked at your correction, it seems worth tweaking a little more:
```diff
-	if (!sbody->GetMass() && !b->GetPhysRadius()) {
-		return luaL_error(l, "the target body has zero mass and physical radius");
+	if (!sbody->GetMass() || !b->GetPhysRadius()) {
+		return luaL_error(l, "the target body has zero mass or zero physical radius");
```
Otherwise, for example, a body with zero mass but non-zero radius (an orbital) will not be discarded. (If I'm not mistaken and the orbital has zero mass, otherwise we'll have to do some more reworking.)

Comment (Web-eWorks) [reply src/lua/LuaSpace.cpp]: I'd recommend separating this into two separate `if` blocks, testing individually for `sbody->GetMass() == 0` and `sbody->GetPhysRadius() == 0`. Both parameters are required to be non-zero to generate a valid orbit.

Comment (Max5377) [reply src/lua/LuaSpace.cpp]: @Web-eWorks Did you approach with a little tweaking by borrowing physical radius from body itself, because SystemBody doesn't have physical radius.

**Review by Gliese852 — APPROVED** [review]

**Review by Web-eWorks — APPROVED** [review]

**Pull request merged**
